near-lite-client/src/block_validation.rs

```diff
@@ -1,146 +1,134 @@
-            continue;
-        }
-
-        approved_stake += bp_stake;
-
-        let validator_public_key = bp_stake_view.public_key.clone();
-        if !maybe_signature
-            .as_ref()
-            .unwrap()
-            .verify(&approval_message, validator_public_key.clone())
-        {
-            return Ok(false);
-        }
-    }
-
-    let threshold = total_stake * 2 / 3;
-    if approved_stake <= threshold {
-        return Ok(false);
-    }
-
-    // # (6)
-    if block_view.next_bps.is_some() {
-        let block_view_next_bps_serialized =
-            block_view.next_bps.as_deref().unwrap().try_to_vec()?;
-        if D::digest(block_view_next_bps_serialized).as_slice()
-            != block_view.inner_lite.next_bp_hash.as_ref()
-        {
-            return Ok(false);
-        }
-    }
-    Ok(true)
+
+pub fn validate_light_block<H: HostFunctions>(
+	head: &LightClientBlockView,
+	block_view: &LightClientBlockView,
+	epoch_block_producers_map: &BTreeMap<CryptoHash, Vec<ValidatorStakeView>>,
+) -> LiteClientResult<()> {
+	//The light client updates its head with the information from LightClientBlockView iff:
+
+	// 1. The height of the block is higher than the height of the current head;
+	// 2. The epoch of the block is equal to the epoch_id or next_epoch_id known for the current
+	// head; 3. If the epoch of the block is equal to the next_epoch_id of the head, then next_bps
+	// is not None; 4. approvals_after_next contain valid signatures on approval_message from the
+	// block producers of the corresponding epoch
+	// 5. The signatures present in approvals_after_next correspond to more than 2/3 of the total
+	// stake (see next section). 6. If next_bps is not none, sha256(borsh(next_bps)) corresponds to
+	// the next_bp_hash in inner_lite.
+
+	// QUESTION: do we also want to pass the block hash received from the RPC?
+	// it's not on the spec, but it's an extra validation
+	let (_current_block_hash, _next_block_hash, approval_message) =
+		reconstruct_light_client_block_view_fields::<H>(block_view)?;
+
+	// (1)
+	if block_view.inner_lite.height <= head.inner_lite.height {
+		return Err(NearLiteClientError::InvalidLiteBlock(String::from(
+			"block view height is not ahead of the head's height",
+		)));
+	}
+
+	// (2)
+	if ![head.inner_lite.epoch_id, head.inner_lite.next_epoch_id]
+		.contains(&block_view.inner_lite.epoch_id)
+	{
+		return Err(NearLiteClientError::InvalidLiteBlock(String::from(
+			"block view epoch id not present in the head",
+		)));
+	}
+
+	// (3)
+	if block_view.inner_lite.epoch_id == head.inner_lite.next_epoch_id
+		&& block_view.next_bps.is_none()
+	{
+		return Err(NearLiteClientError::InvalidLiteBlock(String::from(
+			"block view epoch id is not the next epoch",
+		)));
+	}
+
+	//  (4) and (5)
+	let mut total_stake = 0;
+	let mut approved_stake = 0;
+
+	let epoch_block_producers = &epoch_block_producers_map[&block_view.inner_lite.epoch_id];
+
+	for (maybe_signature, block_producer) in
```

Comment:
Is it possible we check if the total stake reaches the threshold before verifying signatures, since signature verification is expensive?

Reply:
I thought about it again, and I'm not sure of this. 
- If the total staked is enough, the check is not necessary. We're still validating it after signature verification.
- if the total staked is insufficient, the check will be valid.
Do we want to increase the gas cost (adding extra verification) to those submitting correct blocks? The relayer can certainly verify this off-chain before submitting it, and avoid having to pay for a tx that will fail.
@seunlanlege @Wizdave97 @vmarkushin

Reply:
I see what you mean, we only add the approved stake if the signature is verified, so we can't optimize here